onnxruntime/core/providers/webnn/builders/impl/builder_utils.cc

```diff
@@ -68,5 +68,93 @@ common::Status HandleAutoPad(const std::vector<int64_t> input_shape,
   return Status::OK();
 }
 
+common::Status ComputeConvTransposePadAndOutputShape(
+    const int64_t in_size,
+    const int64_t stride,
+    const int64_t kernel,
+    const int64_t dilation,
+    const int64_t adj,
+    AutoPadType pad_type,
+    int64_t& pad_head,
+    int64_t& pad_tail,
+    int64_t& out_size) {
+  // Output shape is explicitly provided - pad values will have to be computed.
+  if (out_size != -1) {
+    // total pad
+    auto total_pad = ComputeTotalPad(in_size, stride, adj, kernel, dilation, out_size);
+    DistributePadding(pad_type, total_pad, pad_head, pad_tail);
+    return Status::OK();
+  }
+
+  // Output shape is not provided - it needs to be computed along with pad values (if applicable).
+
+  // Compute padding if the auto_pad attribute is SAME_UPPER/SAME_LOWER.
+  if (pad_type == AutoPadType::SAME_UPPER || pad_type == AutoPadType::SAME_LOWER) {
+    // The ONNX spec says if `auto_pad` attribute is set, pad until the `out_size`
+    // is `in_size * stride`.
+    auto total_pad = ComputeTotalPad(in_size, stride, adj,
+                                     kernel, dilation, /*out_size = */ in_size * stride);
+    DistributePadding(pad_type, total_pad, pad_head, pad_tail);
+  }
+
+  out_size = (in_size - 1) * stride + adj + (kernel - 1) * dilation + 1 - pad_head - pad_tail;
+
+  return Status::OK();
+}
+
+common::Status ComputeConvTransposePadsAndOutputShape(const std::vector<int64_t> input_shape,
+                                                      const int64_t weight_size_y,
+                                                      const int64_t weight_size_x,
+                                                      const std::vector<int64_t>& onnx_pads,
+                                                      const std::vector<int64_t>& onnx_strides,
+                                                      const std::vector<int64_t>& onnx_dilations,
+                                                      const std::vector<int64_t>& onnx_output_padding,
+                                                      AutoPadType auto_pad_type,
+                                                      std::vector<int64_t>& pads_out,
+                                                      std::vector<int64_t>& output_shape_out,
+                                                      bool use_nchw) {
+  const int64_t input_size_y = use_nchw ? input_shape[2] : input_shape[1];
+  const int64_t input_size_x = use_nchw ? input_shape[3] : input_shape[2];
```

Comment:
Does this logic handle 3D inputs correctly? It's accessing index [3] without checking the size. That's why DML supports rank 3 (https://learn.microsoft.com/en-us/windows/win32/api/directml/ns-directml-dml_convolution_operator_desc) because 3D convolutions were encountered.

Reply:
Also 5D can happen, but that's a separate concern, since WebNN only has conv2d, and so we'd have to decide how to handle that. A flattening reshape of leading dimensions could handle it. Actually, we could reshape 3D up to 4D (appending a trailing dimension of 1 for W or inserting a dimension 1 for H) to potentially simplify the logic here.

Reply:
For 3D inputs, we've ensured it's 4D by appending 1's before calling this method.

For 5D inputs, it's currently fallbacked, it's a good topic to follow up to support 5D inputs for WebNN EP.

Reply:
> For 3D inputs, we've ensured it's 4D by appending 1's before calling this method.

👍

> For 5D inputs, it's currently fallbacked, it's a good topic to follow up to support 5D inputs for WebNN EP.

👍